gapic/src/main/com/google/gapid/util/Trees.java

```diff
@@ -15,112 +15,80 @@
+    TreeItem top = tree.getTopItem();
+    if (top == null) {
       // Work around bug where getTopItem() returns null when scrolling
       // up past the top item (elastic scroll).
-      return null;
+      return (tree.getItemCount() != 0) ? null : emptySet();
     }
-    int treeBottom = rect.y + rect.height;
-    for (TreeItem item = tree.getTopItem(); item != null; item = nextItem(item)) {
-      visible.add(item);
-      Rectangle itemRect = item.getBounds();
-      if (itemRect.y + itemRect.height > treeBottom) {
-        break;
-      }
-    }
-    return visible;
-  }
 
-  /**
-   * @return the i'th child {@link TreeItem} of the {@link Tree} or {@link TreeItem}.
-   */
-  public static TreeItem getItem(Object treeOrTreeItem, int index) {
-    if (treeOrTreeItem instanceof Tree) {
-      return ((Tree)treeOrTreeItem).getItem(index);
+    Rectangle bounds = tree.getClientArea();
+    int treeBottom = bounds.y + bounds.height;
+    Set<TreeItem> visible = Sets.newIdentityHashSet();
+    if (!getVisibleItems(top, treeBottom, visible)) {
+      do {
+        top = getVisibleSiblings(top, treeBottom, visible);
+      } while (top != null);
     }
-    return ((TreeItem)treeOrTreeItem).getItem(index);
+    return visible;
   }
 
   /**
-   * @return the index of the child {@link TreeItem} of the {@link Tree} or {@link TreeItem}.
+   * @return whether the bottom has been reached.
    */
-  public static int indexOf(Object treeOrTreeItem, TreeItem child) {
-    if (treeOrTreeItem instanceof Tree) {
-      return ((Tree)treeOrTreeItem).indexOf(child);
+  private static boolean getVisibleItems(TreeItem item, int treeBottom, Set<TreeItem> visible) {
+    visible.add(item);
+    Rectangle bounds = item.getBounds();
+    if (bounds.y + bounds.height > treeBottom) {
+      return true;
     }
-    return ((TreeItem)treeOrTreeItem).indexOf(child);
-  }
 
-  /**
-   * @return the number of child {@link TreeItem}s of the {@link Tree} or {@link TreeItem}.
-   */
-  public static int itemCount(Object treeOrTreeItem) {
-    if (treeOrTreeItem instanceof Tree) {
-      return ((Tree)treeOrTreeItem).getItemCount();
+    if (item.getExpanded()) {
+      for (TreeItem child : item.getItems()) {
+        if (getVisibleItems(child, treeBottom, visible)) {
+          return true;
+        }
+      }
     }
-    return ((TreeItem)treeOrTreeItem).getItemCount();
+    return false;
   }
 
   /**
-   * @return the parent {@link Tree} or {@link TreeItem} of the {@link TreeItem}.
+   * @return the next parent or {@code null} if the bottom has been reached.
    */
-  public static Object getParent(Object treeOrTreeItem) {
-    if (!(treeOrTreeItem instanceof TreeItem)) {
-      return null;
+  private static TreeItem getVisibleSiblings(TreeItem item, int treeBottom, Set<TreeItem> visible) {
+    TreeItem parent = item.getParentItem();
+    TreeItem[] siblings = (parent == null) ? item.getParent().getItems() : parent.getItems();
```

Comment:
Fold this function into the caller above to avoid repeating these lines for each item.